druid/src/widget/focus.rs

```diff
@@ -0,0 +1,143 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A focus widget.
+
+use druid::widget::prelude::*;
+
+use druid::{
+    commands, Command, Data, FocusNode, HotKey, KbKey, Point, Rect, SysMods, Widget, WidgetPod,
+};
+
+/// A widget that manages a FocusNode to allow focus to be given to this widget and its descendants.
+pub struct Focus<T> {
+    child: WidgetPod<T, Box<dyn Widget<T>>>,
+    auto_focus: bool,
+    focus_requested: bool,
+    focus_node: FocusNode,
+}
+
+impl<T: Data> Focus<T> {
+    /// Create a new Focus widget with a child
+    pub fn new(child: impl Widget<T> + 'static) -> Self {
+        Focus {
+            child: WidgetPod::new(child).boxed(),
+            focus_node: FocusNode::empty(),
+            auto_focus: false,
+            focus_requested: false,
+        }
+    }
+
+    /// Builder-style method to set the `Focus`'s auto focus.
+    pub fn with_auto_focus(mut self, auto_focus: bool) -> Self {
+        self.auto_focus = auto_focus;
+        self
+    }
+}
+
+impl<T: Data> Widget<T> for Focus<T> {
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut T, env: &Env) {
+        let previous_focus_node = ctx.focus_node();
+        ctx.set_focus_node(self.focus_node);
+
+        if self.auto_focus && !self.focus_requested {
+            self.focus_requested = true;
+            ctx.request_focus();
+        }
+
+        match event {
+            Event::MouseDown(_) => {
+                ctx.request_focus();
+                ctx.request_paint();
+            }
+            Event::KeyDown(key_event) => {
+                match key_event {
+                    // Tab and shift+tab
+                    k_e if HotKey::new(None, KbKey::Tab).matches(k_e) => {
+                        ctx.focus_next();
+                    }
+                    k_e if HotKey::new(SysMods::Shift, KbKey::Tab).matches(k_e) => {
+                        ctx.focus_prev();
+                    }
+                    _ => (),
+                };
+
+                ctx.request_paint();
+            }
+            Event::Command(cmd) if cmd.is(commands::REQUEST_FOCUS) => {
+                let widget_id = *cmd.get_unchecked(commands::REQUEST_FOCUS);
+
+                if widget_id == ctx.widget_id() {
+                    ctx.request_focus();
+                    ctx.request_paint();
+                }
+            }
+            _ => (),
+        }
+
+        self.child.event(ctx, event, data, env);
+        ctx.set_focus_node(previous_focus_node);
+    }
+
+    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle, data: &T, env: &Env) {
+        let previous_focus_node = ctx.focus_node();
+
+        match event {
+            LifeCycle::WidgetAdded => {
+                self.focus_node.widget_id = Some(ctx.widget_id());
+                ctx.set_focus_node(self.focus_node);
+                ctx.register_for_focus();
```

Comment:
I'm a little concerned about how complicated this seems to be, although I supposed it's fine if druid users never need to do this themselves, and they just use the `Focus` widget if they want to participate in focus?

Reply:
I think druid users should never do this themselves; they should use the `Focus` widget if they would like to manage focus. After this PR, I wanted to remove focus management from the `TextBox` widget. I've done this already in my pet project: https://github.com/tech-paws/druid-frontend/blob/master/src/ui/widgets/textbox.rs But I think in druid it should be a little bit different.